### PR TITLE
Hold streams together for faster plot callbacks

### PR DESCRIPTION
**Pull request by ahuang11**
Philipp suggested this; not sure if the following is necessary:
```
                if self.plot.comm:
                    push_on_root(self.plot.root.ref['id'])
```

### CODE REVIEW

**Comment by ahuang11** [comment]
I tried this too, but I'm not sure why `self.plot.root` is a `Row` instead of a bokeh `figure`
```python
            with self.plot.root.hold_render():
                with set_curdoc(self.plot.document):
                    Stream.trigger(streams)
```

**Comment by philippjfr** [comment]
>but I'm not sure why self.plot.root is a Row instead of a bokeh figure

The root is the Bokeh root on the Bokeh document and will change depending on whether you are rendering the plot on its own or inside some Panel component.

**Comment by philippjfr** [comment]
I'd suggest using:

```
            with self.plot.state.hold_render():
```

File: holoviews/plotting/bokeh/callbacks.py [hunk 1]
```diff
@@ -225,8 +225,12 @@ def on_msg(self, msg):
             streams.append(stream)
 
         try:
+            self.plot.document.hold()
             with set_curdoc(self.plot.document):
                 Stream.trigger(streams)
+                if self.plot.comm:
+                    push_on_root(self.plot.root.ref['id'])
+            self.plot.document.unhold()
```
Comment (ahuang11) [inline holoviews/plotting/bokeh/callbacks.py:233]:
```suggestion
        with self.plot.state.hold_render():
            with set_curdoc(self.plot.document):
                Stream.trigger(streams)
```

File: holoviews/plotting/bokeh/callbacks.py [hunk 2]
```diff
@@ -225,6 +225,7 @@ def on_msg(self, msg):
             streams.append(stream)
 
         try:
+        with self.plot.state.hold_render():
             with set_curdoc(self.plot.document):
                 Stream.trigger(streams)
```
Comment (ahuang11) [inline holoviews/plotting/bokeh/callbacks.py:230]:
```suggestion
            with self.plot.state.hold_render():
                with set_curdoc(self.plot.document):
                    Stream.trigger(streams)
```

File: holoviews/plotting/bokeh/callbacks.py [hunk 3]
```diff
@@ -225,8 +225,9 @@ def on_msg(self, msg):
             streams.append(stream)
 
         try:
-            with set_curdoc(self.plot.document):
-                Stream.trigger(streams)
+            with self.plot.state.hold_render():
+                with set_curdoc(self.plot.document):
+                    Stream.trigger(streams)
```
Comment (philippjfr) [inline holoviews/plotting/bokeh/callbacks.py:230]:
```suggestion
                with set_curdoc(self.plot.document):
                    self.plot.state.hold_render = True
                    try:
                        Stream.trigger(streams)
                    finally:
                        self.plot.state.hold_render = False
```

File: holoviews/plotting/bokeh/callbacks.py [hunk 4]
```diff
@@ -225,8 +225,12 @@ def on_msg(self, msg):
...
 
         try:
-            with set_curdoc(self.plot.document):
-                Stream.trigger(streams)
+                with set_curdoc(self.plot.document):
+                    self.plot.state.hold_render = True
+                    try:
+                        Stream.trigger(streams)
+                    finally:
+                        self.plot.state.hold_render = False
```
Comment (philippjfr) [inline holoviews/plotting/bokeh/callbacks.py:233]:
```suggestion
            with set_curdoc(self.plot.document):
                self.plot.state.hold_render = True
                try:
                    Stream.trigger(streams)
                finally:
                    self.plot.state.hold_render = False
```

**Comment by philippjfr** [comment]
@ahuang11 Can you confirm whether this still gives the expected speedups?

**Comment by ahuang11** [comment]
It seems so.

This is `self.plot.state.hold_render`(A)

https://github.com/holoviz/holoviews/assets/15331990/6fc4ab3c-6d7e-4b69-bb5d-489ecd01e0a4

And this is `self.plot.document.hold()` (B)

https://github.com/holoviz/holoviews/assets/15331990/c5f8f8e4-247f-4be5-a410-3c52ec63ea4c

`hold_render` seems to do it in batches, and potentially seem a little faster because upon wheel zoom, it throttles vs non-throttle. I do like how (B) updates each line individually for a sense of progress.

**Comment by ahuang11** [comment]
Seems like it works in notebook (VSCode)

https://github.com/holoviz/holoviews/assets/15331990/863935b6-27e1-46d9-9bc2-539ad16eaa1f

**Comment by philippjfr** [comment]
That looks much slower than what we saw originally, no?

**Comment by droumis** [comment]
Seeing 2-4 second lag on my end for render updates. I'm also including in here Philipp's [fix](https://github.com/holoviz/holoviews/pull/6256) to the rangetoolink to work with subcoordinate_y. 

One point to note is that I'm not subselecting channels based on the viewport, just time slicing. Here's the version of the [code](https://github.com/holoviz-topics/neuro/blob/2799cfa5aad95ed63fd2374d2d1e1862416f6218/workflows/multi_channel_timeseries/large_multi-chan-ts.ipynb) I used for this.

https://github.com/holoviz/holoviews/assets/6613202/14000a09-e4f2-4d5c-ab4c-72c4d091411c


What was the code block we initially saw very good speed with? I'll try that again too

**Comment by droumis** [comment]
Assuming the 'original' code block was the following

```python
                self.plot.document.hold()
                try:
                    Stream.trigger(streams)
                finally:
                    self.plot.document.unhold()
```

With the original approach, I'm seeing slightly faster times, ~1-3 seconds for updates:

https://github.com/holoviz/holoviews/assets/6613202/fd37fdd8-1725-4b67-820c-23dcc66652e2

**Comment by philippjfr** [comment]
Someone should dig into this with the browser performance profiler like we did last time to see what it's doing while you wait for the update.

**Comment by ahuang11** [comment]
I'd define channel_slice like this:

```

    channel_slice = slice(y_range[0], y_range[1])

    # calculate the appropriate pyramid level and


    ...
    # extract new data and re-paint the plot
    ds = _extract_ds(ts_dt, pyramid_level, channels).sel(time=time_slice, channel=channel_slice).load()

    ...
    for channel in ds["channel"].values.tolist():
```

**Comment by droumis** [comment]
Ok, I've added the channel slicing. The performance is better now all around, even without holding the render. But now I'm seeing a lot of fragility with the channel y-ticks while navigating around.. e.g.:

<details> <summary> bad channel yticks when channel_slicing </summary>

image

</details>

**Comment by droumis** [comment]
# Initial Profiling:
<details> <summary> Baseline (with channel slicing, no render holding): </summary>

https://github.com/holoviz/holoviews/assets/6613202/7bf743e0-03be-40d3-8fe4-d88f0384ad19

</details>

<details> <summary> plot.document.hold </summary>

https://github.com/holoviz/holoviews/assets/6613202/82307a8c-b728-4b81-ba97-25d2f836bff5

</details>

<details> <summary> plot.state.hold_render </summary>

Sometimes can take very long to update, other times very quick

https://github.com/holoviz/holoviews/assets/6613202/99f88b9f-a45f-4b0f-9d1d-c565e23a9ba1

</details>

**Comment by ahuang11** [comment]
> @ahuang11 , do you see this as well? Is it the same thing as mentioned https://github.com/holoviz/holoviews/pull/6256#issuecomment-2150258611?

Yes that's the same issue.

Thanks for doing the profiling!

**Comment by philippjfr** [comment]
Closing in favor of https://github.com/holoviz/holoviews/pull/6265

**Pull request closed without merging**